File: PHP/cades-signature-server-key.php
<?php

/*
 * This file perform a local CAdES signature in one step using PKI Express.
 */

require __DIR__ . '/vendor/autoload.php';

use Lacuna\PkiExpress\StandardSignaturePolicies;
use Lacuna\PkiExpress\CadesSigner;

// Retrieve the URL argument "userfile", this is filled after been redirected here by the file
// upload.php. We'll set the path of the file to be signed, which was saved in the "app-data" folder
// by upload.php.
$userfile = isset($_GET['userfile']) ? $_GET['userfile'] : null;

try {

    // Verify if the provided userfile exists.
    if (!file_exists("app-data/$userfile")) {
        throw new \Exception('File not found!');
    }

    // Get an instance of the CadesSigner class, responsible for receiving the signature elements
    // and performing the local signature.
    $signer = new CadesSigner();

    // Set PKI default options (see Util.php).
    setPkiDefaults($signer);

    // Set signature policy.
    $signer->signaturePolicy = StandardSignaturePolicies::PKI_BRAZIL_CADES_ADR_BASICA;

    // Set file to be signed. If the file is a CMS, the PKI Express will recognize that and will
    // co-sign that file.
    $signer->setFileToSign("app-data/{$userfile}");

    // The PKCS #12 certificate path.
    $signer->setPkcs12("content/Pierre de Fermat.pfx");
    // Set the certificate's PIN.
    $signer->setCertPassword("1234");

    // Set 'encapsulate content' option (default: true).
    $signer->encapsulateContent = true;

    // Generate path for output file and add to signer object.
    createAppData(); // make sure the "app-data" folder exists (util.php)'
    $outputFile = uniqid() . ".p7s";
    $signer->setOutputFile("app-data/{$outputFile}");

    // Get the file's extension to be passed to signature package generation.
    $ext = pathinfo($userfile, PATHINFO_EXTENSION);

    // Perform the signature.
    $signer->sign();

} catch (Exception $e) {

    // Get exception message to be rendered on signature page.
    $errorMessage = $e->getMessage();
}

?>

<!DOCTYPE html>
<html>
<head>
    <title>CAdES Signature with a server key</title>
    <?php include 'includes.php' // jQuery and other libs (used only to provide a better user experience, but NOT required to use the Web PKI component). ?>
</head>
<body>

<?php include 'menu.php' // The top menu, this can be removed entirely. ?>


<div class="container">

    <?php if (!isset($errorMessage)) { ?>

        <?php
        // If no errors have occurred, this page is shown for the user, with the link to the signed
        // file.
        ?>
        <h2>CAdES Signature with a server key</h2>

        <p>File signed successfully!</p>

        <h3>Actions:</h3>
        <ul>
            <li><a href="app-data/<?= $outputFile ?>">Download the signed file</a></li>
            <li><a href="signature-package.php?file=<?= $outputFile ?>&ext=<?= $ext ?>">Download a signature package of the signed file</a>*</li>
            <li><a href="open-cades-signature.php?userfile=<?= $outputFile ?>">Open/validate the signed file</a></li>
        </ul>

        <p>* This operation requires that the Zip extension to be installed.</p>

    <?php } else { ?>

        <?php
        // If some error occurred, the error message is show with a "Try Again" button to return to
        // the upload.php page.
        ?>
        <div class="alert alert-danger" role="alert" style="margin-top: 2%;">
            <label for="errorMsg">Signature Failed</label><br/>
            <span id="errorMsg"><?= $errorMessage ?></span>
        </div>
        <a class="btn btn-default" href="upload.php?goto=cades-signature-server-key">Try Again</a>

    <?php } ?>

</div>

</body>
</html>

File: Python/sample/views/cades_signature_server_key.py
"""

This sample performs a local CAdES signature in one step usign PKI Express.

"""
import os
import uuid

from flask import Blueprint, render_template, current_app, make_response
from pkiexpress import CadesSigner, standard_signature_policies

from sample.utils import set_pki_defaults, create_app_data, \
    get_expired_page_headers

blueprint = Blueprint('cades_signature_server_key', __name__,
                      url_prefix='/cades-signature-server-key')


@blueprint.route('/<userfile>')
def index(userfile):

    try:

        # Verify if the provided userfile exists.
        if not os.path.exists(os.path.join(current_app.config['APPDATA_FOLDER'],
                                           userfile)):
            return render_template('error.html', msg='File not found')

        # Get an instance of the CadesSigner class, responsible for receiving
        # the signature elements and performing the local signature.
        signer = CadesSigner()

        # Set PKI default options (see utils.py).
        set_pki_defaults(signer)

        # Set signature policy.
        signer.signature_policy = \
            standard_signature_policies.PKI_BRAZIL_CADES_ADR_BASICA

        # Set file to be signed. If the file is a CSM, the PKI Express will
        # recognize that and will co-sign that file.
        signer.set_file_to_sign_from_path(os.path.join(
            current_app.config['APPDATA_FOLDER'], userfile))

        # The PKCS #12 certificate path.
        signer.set_pkcs12_from_path(os.path.join(current_app.static_folder,
                                                 'Pierre de Fermat.pfx'))
        # Set the certificate's PIN.
        signer.cert_password = '1234'

        # Set 'encapsulate content' option (default: True).
        signer.encapsulated_content = True

        # Generate path for output file and add to signer object.
        create_app_data()  # Guarantees that "app_data" folder exists.
        output_file = '%s.p7s' % (str(uuid.uuid4()))
        signer.output_file = os.path.join(current_app.config['APPDATA_FOLDER'],
                                          output_file)

        # Perform the signature.
        signer_cert = signer.sign(get_cert=True)

        response = make_response(render_template(
            'cades_signature_server_key/index.html',
            signer_cert=signer_cert,
            filename=output_file))
        response.headers = get_expired_page_headers()
        return response

    except Exception as e:
        return render_template('error.html', msg=e)